File: hdf-diff/test/matching/vendor-fuzzy-title.test.ts
import { describe, it, expect } from 'vitest';
import {
  createVendorFuzzyTitleStrategy,
  levenshteinDistance,
  normalizedLevenshtein,
} from '../../src/matching/vendor-fuzzy-title.js';
import { autoDetectPrefix, normalizeTitle } from '../../src/matching/vendor-fuzzy-title.js';

describe('levenshteinDistance', () => {
  it('should return 0 for identical strings', () => {
    expect(levenshteinDistance('abc', 'abc')).toBe(0);
  });

  it('should handle empty strings', () => {
    expect(levenshteinDistance('', '')).toBe(0);
    expect(levenshteinDistance('abc', '')).toBe(3);
    expect(levenshteinDistance('', 'abc')).toBe(3);
  });

  it('should compute correct edit distance', () => {
    expect(levenshteinDistance('kitten', 'sitting')).toBe(3);
    expect(levenshteinDistance('Saturday', 'Sunday')).toBe(3);
  });

  it('should handle single character', () => {
    expect(levenshteinDistance('a', 'b')).toBe(1);
    expect(levenshteinDistance('a', 'a')).toBe(0);
  });
});

describe('normalizedLevenshtein', () => {
  it('should return 0.0 for identical strings', () => {
    expect(normalizedLevenshtein('abc', 'abc')).toBe(0.0);
  });

  it('should return 0.0 for both empty', () => {
    expect(normalizedLevenshtein('', '')).toBe(0.0);
  });

  it('should return 1.0 for completely different single chars', () => {
    expect(normalizedLevenshtein('a', 'b')).toBe(1.0);
  });

  it('should return value between 0 and 1', () => {
    const dist = normalizedLevenshtein('kitten', 'sitting');
    expect(dist).toBeGreaterThan(0);
    expect(dist).toBeLessThan(1);
  });
});

describe('autoDetectPrefix', () => {
  it('should detect dominant multi-token prefix', () => {
    const titles = [
      'RHEL 9 must be supported.',
      'RHEL 9 must check GPG signatures.',
      'RHEL 9 must disable audit.',
      'Ubuntu 22 must use TLS.',
    ];
    expect(autoDetectPrefix(titles)).toBe('RHEL 9');
  });

  it('should detect prefix with outliers', () => {
    const titles = [
      'Nutanix VMM must enable logging.',
      'Nutanix VMM must restrict access.',
      'Nutanix VMM must enforce TLS.',
      'Something else must exist.',
    ];
    expect(autoDetectPrefix(titles)).toBe('Nutanix VMM');
  });

  it('should return empty for feature-focused corpus (no dominant prefix)', () => {
    const titles = [
      'Ensure password complexity is set.',
      'Verify SSH is configured.',
      'Check file permissions.',
      'Audit log rotation.',
    ];
    expect(autoDetectPrefix(titles)).toBe('');
  });

  it('should return empty for empty array', () => {
    expect(autoDetectPrefix([])).toBe('');
  });

  it('should stop before modal verbs', () => {
    const titles = [
      'The Apache web server must be configured.',
      'The Apache web server must limit connections.',
      'The Apache web server must use TLS.',
    ];
    expect(autoDetectPrefix(titles)).toBe('The Apache web server');
  });
});

describe('normalizeTitle', () => {
  it('should strip leading prefix', () => {
    expect(normalizeTitle('RHEL 9 must be supported.', 'RHEL 9')).toBe('must be supported.');
  });

  it('should return unchanged when prefix is empty', () => {
    expect(normalizeTitle('must be supported.', '')).toBe('must be supported.');
  });

  it('should return unchanged when prefix not at start', () => {
    expect(normalizeTitle('Ubuntu must use RHEL 9 tools.', 'RHEL 9')).toBe(
      'Ubuntu must use RHEL 9 tools.',
    );
  });

  it('should handle title equal to prefix', () => {
    expect(normalizeTitle('RHEL 9', 'RHEL 9')).toBe('');
  });
});

describe('VendorFuzzyTitleStrategy', () => {
  const strategy = createVendorFuzzyTitleStrategy();

  it('should have name "vendorFuzzyTitle"', () => {
    expect(strategy.name).toBe('vendorFuzzyTitle');
  });

  it('should match cross-vendor titles after prefix normalization', () => {
    const oldReqs = [
      { id: 'V-001', title: 'RHEL 9 must enable audit logging.', impact: 0.7, tags: {} },
      { id: 'V-002', title: 'RHEL 9 must configure SSH.', impact: 0.5, tags: {} },
    ];
    const newReqs = [
      {
        id: 'AL-001',
        title: 'Amazon Linux 2023 must enable audit logging.',
        impact: 0.7,
        tags: {},
      },
      { id: 'AL-002', title: 'Amazon Linux 2023 must configure SSH.', impact: 0.5, tags: {} },
    ];

    const result = strategy.match(oldReqs, newReqs);

    expect(result.matched).toHaveLength(2);
    expect(result.unmatchedOld).toHaveLength(0);
    expect(result.unmatchedNew).toHaveLength(0);
  });

  it('should reject unrelated titles', () => {
    const oldReqs = [
      { id: 'V-001', title: 'RHEL 9 must enable audit logging.', impact: 0.7, tags: {} },
    ];
    const newReqs = [
      { id: 'AL-001', title: 'Amazon Linux 2023 must configure TLS certificates.', impact: 0.7, tags: {} },
    ];

    const result = strategy.match(oldReqs, newReqs);

    expect(result.matched).toHaveLength(0);
    expect(result.unmatchedOld).toHaveLength(1);
    expect(result.unmatchedNew).toHaveLength(1);
  });

  it('should handle requirements with no title', () => {
    const oldReqs = [{ id: 'V-001', impact: 0.7, tags: {} }];
    const newReqs = [{ id: 'AL-001', impact: 0.7, tags: {} }];

    const result = strategy.match(oldReqs, newReqs);

    expect(result.matched).toHaveLength(0);
  });

  it('should handle empty inputs', () => {
    const result = strategy.match([], []);
    expect(result.matched).toHaveLength(0);
  });

  it('should use relationship field', () => {
    const oldReqs = [
      { id: 'V-001', title: 'RHEL 9 must enable audit logging.', impact: 0.7, tags: {} },
    ];
    const newReqs = [
      {
        id: 'AL-001',
        title: 'Amazon Linux 2023 must enable audit logging.',
        impact: 0.7,
        tags: {},
      },
    ];

    const result = strategy.match(oldReqs, newReqs);

    expect(result.matched).toHaveLength(1);
    expect(result.matched[0]!.relationship).toBe('primary');
  });

  it('should work with feature-focused corpus (no prefix to strip)', () => {
    const oldReqs = [
      { id: 'V-001', title: 'Ensure password complexity is set.', impact: 0.7, tags: {} },
      { id: 'V-002', title: 'Verify SSH is configured properly.', impact: 0.5, tags: {} },
    ];
    const newReqs = [
      { id: 'R-001', title: 'Ensure password complexity is set.', impact: 0.7, tags: {} },
      { id: 'R-002', title: 'Verify SSH is configured properly.', impact: 0.5, tags: {} },
    ];

    const result = strategy.match(oldReqs, newReqs);

    expect(result.matched).toHaveLength(2);
  });
});

File: hdf-diff/go/matching/vendor_fuzzy_title_test.go
package matching

import (
	"strings"
	"testing"

	hdf "github.com/mitre/hdf-libs/hdf-schema/dist/go/v3"
	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
)

// TestLevenshteinDistance_BoundsOverlongInput: pathological over-cap inputs are
// not run through the O(m*n) DP — they return the trivial upper bound max(m,n).
// Realistic-length titles keep their exact distance.
func TestLevenshteinDistance_BoundsOverlongInput(t *testing.T) {
	a := strings.Repeat("x", maxLevenshteinRunes+1000)
	b := a + strings.Repeat("y", 100) // exact edit distance is 100
	got := LevenshteinDistance(a, b)
	assert.Equal(t, len([]rune(b)), got,
		"over-cap inputs return max(m,n), not the exact distance")

	// A boundary-length pair (exactly at the cap) is still computed exactly.
	c := strings.Repeat("x", maxLevenshteinRunes)
	d := strings.Repeat("x", maxLevenshteinRunes-1) + "y"
	assert.Equal(t, 1, LevenshteinDistance(c, d), "at-cap inputs keep the exact distance")

	// Normal titles are unaffected.
	assert.Equal(t, 3, LevenshteinDistance("kitten", "sitting"))
}

func TestLevenshteinDistance(t *testing.T) {
	assert.Equal(t, 0, LevenshteinDistance("abc", "abc"))
	assert.Equal(t, 3, LevenshteinDistance("", "abc"))
	assert.Equal(t, 3, LevenshteinDistance("abc", ""))
	assert.Equal(t, 0, LevenshteinDistance("", ""))
	assert.Equal(t, 3, LevenshteinDistance("kitten", "sitting"))
	assert.Equal(t, 3, LevenshteinDistance("Saturday", "Sunday"))
	assert.Equal(t, 1, LevenshteinDistance("a", "b"))
}

func TestNormalizedLevenshtein(t *testing.T) {
	assert.Equal(t, 0.0, NormalizedLevenshtein("abc", "abc"))
	assert.Equal(t, 0.0, NormalizedLevenshtein("", ""))
	assert.Equal(t, 1.0, NormalizedLevenshtein("a", "b"))
	dist := NormalizedLevenshtein("kitten", "sitting")
	assert.Greater(t, dist, 0.0)
	assert.Less(t, dist, 1.0)
}

func TestAutoDetectPrefix(t *testing.T) {
	t.Run("dominant prefix", func(t *testing.T) {
		titles := []string{
			"RHEL 9 must be supported.",
			"RHEL 9 must check GPG signatures.",
			"RHEL 9 must disable audit.",
			"Ubuntu 22 must use TLS.",
		}
		assert.Equal(t, "RHEL 9", AutoDetectPrefix(titles, 0.5))
	})

	t.Run("no dominant prefix", func(t *testing.T) {
		titles := []string{
			"Ensure password complexity is set.",
			"Verify SSH is configured.",
			"Check file permissions.",
			"Audit log rotation.",
		}
		assert.Equal(t, "", AutoDetectPrefix(titles, 0.5))
	})

	t.Run("empty", func(t *testing.T) {
		assert.Equal(t, "", AutoDetectPrefix(nil, 0.5))
	})

	t.Run("stops before modal", func(t *testing.T) {
		titles := []string{
			"The Apache web server must be configured.",
			"The Apache web server must limit connections.",
			"The Apache web server must use TLS.",
		}
		assert.Equal(t, "The Apache web server", AutoDetectPrefix(titles, 0.5))
	})
}

func TestNormalizeTitle(t *testing.T) {
	assert.Equal(t, "must be supported.", NormalizeTitle("RHEL 9 must be supported.", "RHEL 9"))
	assert.Equal(t, "must be supported.", NormalizeTitle("must be supported.", ""))
	assert.Equal(t, "Ubuntu must use RHEL 9 tools.", NormalizeTitle("Ubuntu must use RHEL 9 tools.", "RHEL 9"))
	assert.Equal(t, "", NormalizeTitle("RHEL 9", "RHEL 9"))
}

func TestVendorFuzzyTitleStrategy_Name(t *testing.T) {
	s := NewVendorFuzzyTitleStrategy(0)
	assert.Equal(t, "vendorFuzzyTitle", s.Name())
}

func TestVendorFuzzyTitleStrategy_CrossVendorMatch(t *testing.T) {
	s := NewVendorFuzzyTitleStrategy(0)
	oldReqs := []hdf.EvaluatedRequirement{
		{ID: "V-001", Impact: 0.7, Title: strPtr("RHEL 9 must enable audit logging.")},
		{ID: "V-002", Impact: 0.5, Title: strPtr("RHEL 9 must configure SSH.")},
	}
	newReqs := []hdf.EvaluatedRequirement{
		{ID: "AL-001", Impact: 0.7, Title: strPtr("Amazon Linux 2023 must enable audit logging.")},
		{ID: "AL-002", Impact: 0.5, Title: strPtr("Amazon Linux 2023 must configure SSH.")},
	}

	result := s.Match(oldReqs, newReqs)

	require.Len(t, result.Matched, 2)
	assert.Len(t, result.UnmatchedOld, 0)
	assert.Len(t, result.UnmatchedNew, 0)
}

func TestVendorFuzzyTitleStrategy_RejectUnrelated(t *testing.T) {
	s := NewVendorFuzzyTitleStrategy(0)
	oldReqs := []hdf.EvaluatedRequirement{
		{ID: "V-001", Impact: 0.7, Title: strPtr("RHEL 9 must enable audit logging.")},
	}
	newReqs := []hdf.EvaluatedRequirement{
		{ID: "AL-001", Impact: 0.7, Title: strPtr("Amazon Linux 2023 must configure TLS certificates.")},
	}

	result := s.Match(oldReqs, newReqs)

	assert.Len(t, result.Matched, 0)
	assert.Len(t, result.UnmatchedOld, 1)
	assert.Len(t, result.UnmatchedNew, 1)
}

func TestVendorFuzzyTitleStrategy_NoTitle(t *testing.T) {
	s := NewVendorFuzzyTitleStrategy(0)
	oldReqs := []hdf.EvaluatedRequirement{{ID: "V-001", Impact: 0.7}}
	newReqs := []hdf.EvaluatedRequirement{{ID: "AL-001", Impact: 0.7}}

	result := s.Match(oldReqs, newReqs)

	assert.Len(t, result.Matched, 0)
}

func TestVendorFuzzyTitleStrategy_Empty(t *testing.T) {
	s := NewVendorFuzzyTitleStrategy(0)
	result := s.Match(nil, nil)
	assert.Len(t, result.Matched, 0)
}

func TestVendorFuzzyTitleStrategy_Relationship(t *testing.T) {
	s := NewVendorFuzzyTitleStrategy(0)
	oldReqs := []hdf.EvaluatedRequirement{
		{ID: "V-001", Impact: 0.7, Title: strPtr("RHEL 9 must enable audit logging.")},
	}
	newReqs := []hdf.EvaluatedRequirement{
		{ID: "AL-001", Impact: 0.7, Title: strPtr("Amazon Linux 2023 must enable audit logging.")},
	}

	result := s.Match(oldReqs, newReqs)

	require.Len(t, result.Matched, 1)
	assert.Equal(t, "primary", result.Matched[0].Relationship)
}

func TestVendorFuzzyTitleStrategy_FeatureCorpus(t *testing.T) {
	s := NewVendorFuzzyTitleStrategy(0)
	oldReqs := []hdf.EvaluatedRequirement{
		{ID: "V-001", Impact: 0.7, Title: strPtr("Ensure password complexity is set.")},
		{ID: "V-002", Impact: 0.5, Title: strPtr("Verify SSH is configured properly.")},
	}
	newReqs := []hdf.EvaluatedRequirement{
		{ID: "R-001", Impact: 0.7, Title: strPtr("Ensure password complexity is set.")},
		{ID: "R-002", Impact: 0.5, Title: strPtr("Verify SSH is configured properly.")},
	}

	result := s.Match(oldReqs, newReqs)

	require.Len(t, result.Matched, 2)
}
